src/dev/lupio/lupio_tmr.hh

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2021 The Regents of the University of California
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __DEV_LUPIO_LUPIO_TMR_HH__
+#define __DEV_LUPIO_LUPIO_TMR_HH__
+
+#include "arch/riscv/interrupts.hh"
+#include "dev/io_device.hh"
+#include "dev/platform.hh"
+#include "dev/lupio/lupio_pic.hh"
+#include "params/LupioTMR.hh"
+
+namespace gem5
+{
+
+/**
+ * LupioTMR:
+ * A timer virtual device 
+ */
+class LupioPIC;
+
+class LupioTMR : public BasicPioDevice
+{
+  protected:
+    const ByteOrder byteOrder = ByteOrder::little;
+    System *system;
+    EventFunctionWrapper tmrEvent;
+
+    Tick start = 0; 
+    Tick next = 0; 
+    Cycles startCycle = curCycle(); 
```

Comment:
1. I'd put this in the initializer list with the constructor. I think it's OK to use const initialization in the header, but non const initialization scares me a bit.
2. When restoring from a checkpoint, this isn't going to work. If you really need this to be the current cycle at the beginning of simulation, you should grab it in `startup`. See http://doxygen.gem5.org/release/current/classgem5_1_1SimObject.html#details for details.